tendermint/src/lite/verifier.rs

```diff
@@ -154,3 +165,137 @@ where
+    // middle of the trusted height and the desired height (h2.height).
+
+    let trusted_header = trusted_state.last_header();
+    let trusted_height: u64 = trusted_header.header().height().value();
+    let untrusted_height: u64 = untrusted_header.header().height().value();
+    let pivot: u64 = (trusted_height + untrusted_height) / 2;
+    let pivot_header = req.signed_header(pivot)?;
+    let pivot_next_vals = req.validator_set(pivot + 1)?;
+
+    verify(&pivot_header, &pivot_next_vals)?;
+
+    // Can we trust pivot header based on trusted_state?
+    can_trust(
+        trusted_state,
+        &pivot_header,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+    // Trust the header in between the trusted and (still) untrusted height:
+    store.add(&pivot_header)?;
+    let pivot_trusted = TS::new(pivot_header, pivot_next_vals);
+
+    // Can we trust h2 based on the (now trusted) "pivot header"?
+    can_trust(
+        &pivot_trusted,
+        untrusted_header,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+    // Add to trusted headers:
+    store.add(untrusted_header)?;
+
+    Ok(())
+}
+
+/// This function captures the high level logic of the light client verification, i.e.,
+/// an application call to the light client module to (optionally download) and
+/// verify a header for some height.
+pub fn verify_header<TS, SH, VS, L, S, R>(
+    height: Height,
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = VS>,
+    L: TrustThreshold,
+    S: Store<SignedHeader = SH>,
+    R: Requester<SignedHeader = SH, ValidatorSet = VS>,
+    VS: ValidatorSet,
+    SH: SignedHeader,
+{
+    // Check if we already trusted a header at the given height and it didn't expire:
+    if let Ok(sh2) = store.get(height) {
+        is_within_trust_period(sh2.header(), trusting_period, now)?
+    }
+
+    // We haven't trusted a header at height yet. Request it:
+    let sh2 = req.signed_header(height)?;
+    let sh2_next_vals = req.validator_set(height.increment())?;
+    verify(&sh2, &sh2_next_vals)?;
+    is_within_trust_period(sh2.header(), trusting_period, now)?;
+
+    // Get the highest trusted header with height lower than sh2's.
+    let sh1 = store.get_smaller_or_equal(height)?;
+    // TODO(Liamsi): it's odd that we need re-request the validator set here; shouldn't this
+    // be stored instead?!
+    let sh1_next_vals = req.validator_set(sh1.header().height().value() + 1)?;
+    let sh1_trusted = TS::new(sh1, sh1_next_vals);
+
+    can_trust(
+        &sh1_trusted,
+        &sh2,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+
+    is_within_trust_period(sh2.header(), trusting_period, now)?;
+    store.add(&sh2)?;
```

Comment:
This is also redundant as `can_trust` also stores the header (or trusted state) if it can be trusted...